### PR TITLE
Hangouts/meeting feature

### LINKED ISSUE #45: Add "online hangout/meeting" feature
We would like to foster connections between community members who may be in different locations around the world.To support this goal, we can add a `Hangout` feature (or something similar like `Meeting`.)

## Task

- [ ] create a new app for hangouts/meetings
- [ ] create a `Hangout` or `Meeting` model (or something similar)
   - [ ] `title` field
   - [ ] `description` RichText field
   - [ ] `topics` keword select field, e.g. with `django-taggit`
   - [ ] `URL` field (optional) to link to external meeting
- [ ] add templates for creating and displaying hangout/meetings
- [ ] in the template, render the meeting title, description,and keywords
- [ ] keywords should link to a keyword page listing related meeting/hangouts
- [ ] if possible, we can render a Jitsi Meet room with an ID based on the database ID

### PR DESCRIPTION
**Pull request by Upasanadhameliya**
Closes #45

### CODE REVIEW

**Comment by Upasanadhameliya** [comment]
@brylie there is a bug that I am facing. I have created `hangouts index page` which is the parent of `hangouts` page. Each hangouts page has a `url` which is a `streamfield panel` of a `url block`. Whenever I create a new `hangouts` page having a url, it displays the following sort of an output, which seems out of the ordinary. Do you have an idea of what might be causing the problem?
ques79
ques80

Notice the page named `old` having a different sort of text in `STATUS`

File: project/hangouts/models.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+from wagtail.admin.edit_handlers import FieldPanel, StreamFieldPanel
+from wagtail.core.fields import RichTextField, StreamField
+from wagtail.core.models import Page
+from wagtail.core.blocks import URLBlock
+
+
+# Create your models here.
+class Hangout(Page):
+    description = RichTextField()
+    topics = TaggableManager()
```
Comment (brylie) [inline project/hangouts/models.py:12]: Yeah, "topics" sounds better than "keywords".

File: project/hangouts/models.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    url = StreamField(
+        [
+            (
+                "hangout_link",
+                URLBlock(required=False, form_classname="hangout link"),
```
Comment (brylie) [inline project/hangouts/models.py:17]: We should consider making "hangout link" a top-level field, rather than embedded in a stream field. Stream fields are good for allowing users to build block-based content. The trade-off is that we hand over content layout to the content editor instead of being able to define specific layouts in an overall Template. E.g. we couldn't define a template that *always* places the hangout link directly below the description if we embed the hangout link in a stream field.

**Review by brylie — CHANGES_REQUESTED** [review]

File: project/hangouts/templates/hangouts/hangouts_index_page.html [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+{{ page.introduction | richtext }}
+
+<div class="row mt-5">
+    {% for hangout in hangouts %}
+    <div class="col">
+        <div class="card mb-5" style="width: 18rem;">
+            <div class="card-body">
+                <h5 class="card-title">{{ hangout.title }}</h5>
+                {{ hangout.description | richtext }}
```
Comment (brylie) [inline project/hangouts/templates/hangouts/hangouts_index_page.html:18]: We may want to control the amount of description text that appears in a Hangout card, e.g. by using the [`truncatewords`](https://docs.djangoproject.com/en/3.2/ref/templates/builtins/#truncatewords) template tag here.

**Review by brylie — CHANGES_REQUESTED** [review]

File: project/hangouts/models.py (hunk 2 above)
Comment (Upasanadhameliya) [reply project/hangouts/models.py:17]: @brylie I am unable to add `hangout_link` as `models.URLField()`. When I do so it gives the following sort of an error

<details>
<summary>Error </summary>

```
Internal Server Error: /admin/pages/add/hangouts/hangout/4/
Traceback (most recent call last):
  File "C:\Users\Upasana Dhameliya\.conda\envs\wagtail38\lib\site-packages\django\core\handlers\exception.py", line 47, in inner
    response = get_response(request)
  File "C:\Users\Upasana Dhameliya\.conda\envs\wagtail38\lib\site-packages\django\core\handlers\base.py", line 181, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "C:\Users\Upasana Dhameliya\.conda\envs\wagtail38\lib\site-packages\django\views\decorators\cache.py", line 44, in _wrapped_view_func
    response = view_func(request, *args, **kwargs)
  File "C:\Users\Upasana Dhameliya\.conda\envs\wagtail38\lib\site-packages\wagtail\admin\urls\__init__.py", line 108, in wrapper
    return view_func(request, *args, **kwargs)
  File "C:\Users\Upasana Dhameliya\.conda\envs\wagtail38\lib\site-packages\wagtail\admin\auth.py", line 193, in decorated_view
    return view_func(request, *args, **kwargs)
  File "C:\Users\Upasana Dhameliya\.conda\envs\wagtail38\lib\site-packages\django\views\generic\base.py", line 70, in view
    return self.dispatch(request, *args, **kwargs)
  File "C:\Users\Upasana Dhameliya\.conda\envs\wagtail38\lib\site-packages\wagtail\admin\views\pages\create.py", line 88, in dispatch
    self.page = self.page_class(owner=self.request.user)
  File "C:\Users\Upasana Dhameliya\.conda\envs\wagtail38\lib\site-packages\wagtail\core\models.py", line 897, in __init__
    super().__init__(*args, **kwargs)
  File "C:\Users\Upasana Dhameliya\.conda\envs\wagtail38\lib\site-packages\modelcluster\models.py", line 170, in __init__
    super(ClusterableModel, self).__init__(*args, **kwargs)
  File "C:\Users\Upasana Dhameliya\.conda\envs\wagtail38\lib\site-packages\django\db\models\base.py", line 488, in __init__
    _setattr(self, field.attname, val)
AttributeError: can't set attribute
```


</details>

Comment (brylie) [reply project/hangouts/models.py:17]: Try deleting the content or local development database before making migrations.

Comment (Upasanadhameliya) [reply project/hangouts/models.py:17]: I deleted the `migrations` of the hangouts app and even the `db.sqlite3`. I ran `makemigrations` and `migrate` anew, yet it gives the same error as above. The `Hangouts` model:

```
class Hangout(Page):
    description = RichTextField()
    topics = TaggableManager()
    url = models.URLField(blank=True)
    content_panels = Page.content_panels + [
        FieldPanel("description",classname="full"),
        FieldPanel("topics",classname="full"),
        FieldPanel("url",classname="full")
    ]
    parent_page_types = [
        "hangouts.HangoutsIndexPage",
    ]
```

Comment (Upasanadhameliya) [reply project/hangouts/models.py:17]: The error above is given when I try to add a child page to the `Hangouts Index Page`.

Comment (brylie) [reply project/hangouts/models.py:17]: Go ahead and push your latest changes so I can help troubleshoot.

Comment (Upasanadhameliya) [reply project/hangouts/models.py]: @brylie I fixed the issue. The problem was caused by naming the `URLField()` as `url`. (See [this](https://github.com/wagtail/wagtail/issues/1710))

**Comment by brylie** [comment]
OK, the PR code should work now. I reset the migrations and flattened the Hangout model hierarchy. I also needed to add some safety checks around our `<Model>.objects.get()` methods since they will throw an `ObjectDoesNotExist` error when no model object exists.

File: project/hangouts/models.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    max_count = 1
+
+    def get_context(self, request, *args, **kwargs):
+        context = super().get_context(request, *args, **kwargs)
+        context["hangouts"] = (
+            self.allowed_subpage_models()[0]
+            .objects.child_of(self)
+            .live()
+            .order_by("title")
+        )
```
Comment (brylie) [inline project/hangouts/models.py:48]: Ensure that this line still works when no hangouts have been created. E.g., we may need to add a `try`/ `catch` in the code, and a conditional in the related template for the case when the hangouts index page exists but no hangouts have been created.

Comment (Upasanadhameliya) [reply project/hangouts/models.py:57]: It still works when no hangouts have been created without the need for `try`/`except`

**Comment by brylie** [comment]
One thing to consider is that, by default, Codebuddies hangouts use Jitsi Meet. The `hangout_url` field would be optional and an alternative to Jitsi Meet. So, we might think about telling users the default hangout uses Jitsi Meet unless they specify an alternative hangout URL. Likewise, the template should show an embedded Jitsi Meet room unless the `hangout_url` is set. In the latter case, we can create a large button like "join the hangout."

File: project/hangouts/models.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        if topic:
+            hangouts = hangouts.filter(topics=topic)
+        return hangouts
+
+    def get_child_topics(self):
+        topics = []
+        for hangout in self.get_hangouts():
+            topics += hangout.get_topics
+        topics = sorted(set(topics))
+        return topics
```
Comment (brylie) [inline project/hangouts/models.py:98]: This helper method should not be needed since we already have a `topics` field.

Comment (brylie) [inline project/hangouts/models.py:98]: This method might be very slow since it is iterating over all the hangouts that have ever been created.

A more performant approach might be to use the [`dictinct`](https://docs.djangoproject.com/en/3.2/ref/models/querysets/#distinct) QuerySet operator on the `HangoutTag` model.

```py
# Assuming each HangoutTag gets a "name" property from TaggedItemBase
hangout_tags = HangoutTag.objects.distinct("name")
```

**Review by brylie — CHANGES_REQUESTED** [review]

File: project/hangouts/models.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+    @property
+    def get_topics(self):
+
+        topics = self.topics.all()
+        for topic in topics:
+            topic.url = "/" + "/".join(
+                s.strip("/") for s in [self.get_parent().url, "topics", topic.slug]
+            )
+        return topics
```
Comment (brylie) [inline project/hangouts/models.py:44]: What is this method (property) used for? The code is a bit hard to follow and is somewhat "low level" since it uses string manipulation. Is there a higher-level approach to whatever problem is intended to be solved here?

**Review by brylie — CHANGES_REQUESTED** [review]

File: project/hangouts/models.py (hunk 6 above)
Comment (Upasanadhameliya) [reply project/hangouts/models.py:44]: This creates a `topic url` based on the `topic slug` in order to filter based on a particular topic. See [this](https://github.com/wagtail/bakerydemo/blob/master/bakerydemo/blog/models.py)

File: project/hangouts/models.py (hunk 5 above)
Comment (brylie) [reply project/hangouts/models.py:98]: The main differences that come to mind with `distinct` and the imperative approach are:

- `distinct` is declarative, so a bit higher-level than an imperative loop
- `distinct` can rely on a database index (on the "name" column, or whichever) to speed up the process

File: project/hangouts/models.py (hunk 6 above)
Comment (brylie) [reply project/hangouts/models.py:44]: At the very least, a docstring comment would be a good start for explaining the context of why this property is useful. From there, we can explore alternatives if needed.

Comment (Upasanadhameliya) [reply project/hangouts/models.py:44]: I will try to figure out some alternative to this code.. meanwhile you can suggest any ideas for the same

Comment (brylie) [reply project/hangouts/models.py:44]: The docstring may be sufficient so long as we understand the use case for the code. 

The idea is to write as little code as is needed, make the code high-level and self-documenting where possible, and add docstring comments to explain the aspects that code can't explain directly.

Comment (brylie) [reply project/hangouts/models.py:44]: I am trying to understand how we will use the method.

The Wagtail Bakery project seems like an excellent reference for many common features made with Wagtail/Django. Thanks for pointing it out.

Comment (Upasanadhameliya) [reply project/hangouts/models.py:44]: Although they have given a disclaimer saying that the project should not be used to start your own site

File: project/hangouts/models.py (hunk 5 above)
Comment (Upasanadhameliya) [reply project/hangouts/models.py:98]: @brylie I am facing a few problems with `distinct`:
- Firstly `distinct` on table rows is only supported by `PostgreSQL` (See [this](https://stackoverflow.com/questions/23959994/django-filter-by-latest-per-distinct-column#:~:text=will%20only%20work%20if%20you%20are%20using%20postgres.)). As a result, I get the following error:
```
(Pdb) self.get_hangouts().distinct('topics')
*** django.db.utils.NotSupportedError: DISTINCT ON fields is not supported by this database backend
```
- Secondly I tried using `distinct` in other ways, but it doesn't work as expected. (Reference [link](https://stackoverflow.com/questions/2466496/select-distinct-values-from-a-table-field#:~:text=To%20query%20your%20DB%2C%20you%20just%20have%20to%20call%3A) )
```
(Pdb) pgqr = self.get_hangouts()

(Pdb) pgqr
<PageQuerySet [<Hangout: Second>, <Hangout: First one>, <Hangout: First>]>

(Pdb) pgqr.distinct()
<PageQuerySet [<Hangout: Second>, <Hangout: First one>, <Hangout: First>]>

(Pdb) pgqr.values('topics')
<PageQuerySet [{'topics': 2}, {'topics': 1}, {'topics': 1}]>

(Pdb) pgqr.values('topics').distinct() #Does not work as expected, wanted to keep {'topics':1} as a single entry 
<PageQuerySet [{'topics': 2}, {'topics': 1}, {'topics': 1}]>

(Pdb) pgqr.values_list('topics').distinct() #This doesn't work either
<PageQuerySet [(2,), (1,), (1,)]>

(Pdb) HangoutTag.objects.distinct() #Neither does this
<QuerySet [<HangoutTag: First tagged with First>, <HangoutTag: First one tagged with First>, <HangoutTag: Second tagged with Second>]>
# In the last case the distinct is considering both relationships in terms of uniqueness
# So it's considering that a "page + tag" should be unique and not just taking unique "tags" itself without the page
```

Comment (Upasanadhameliya) [reply project/hangouts/models.py:98]: I am really stuck on this one and trying to think of ways to change the structure, in a way to help our case

Comment (brylie) [reply project/hangouts/models.py:98]: What about this approach?

https://stackoverflow.com/a/38062244

**Comment by brylie** [comment]
@Upasanadhameliya, I believe this pull request is ready for review, correct?

**Comment by Upasanadhameliya** [comment]
@brylie I don't think so, the templates part is yet to be added

**Comment by brylie** [comment]
All good, no hurry. I'm available now to help as well. Thanks for your help.
